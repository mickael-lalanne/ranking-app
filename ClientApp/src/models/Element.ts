export interface Element {
    id?: string;
    name: string;
    image: string;
    type?: number;
}

export interface RankedElement {
    elementId: string;
    tierId: string;
    position: number;
}
