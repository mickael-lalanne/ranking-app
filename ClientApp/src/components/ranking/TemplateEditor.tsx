import React, { useState, useEffect } from 'react';
import TextField from '@mui/material/TextField';
import TierEditView from './TierEditView';
import { ETemplateMode, TIERS_COLORS, Template, Tier } from '../../models/Template';
import { Element } from '../../models/Element';
import { css } from '@emotion/css';
import ElementEditView from './ElementEditView';
import TmpElementImg from '../../images/tmp_element_img.png';
import IconButton from '@mui/material/IconButton';
import RemoveCircleOutlineIcon from '@mui/icons-material/RemoveCircleOutline';
import DeleteIcon from '@mui/icons-material/Delete';
import AppButton from '../shared/AppButton';
import { useTheme } from '@mui/material/styles';

const TemplateEditor = (
    { saveHandler, template, mode }: {
        saveHandler: (templateToCreate: Template) => Promise<void>,
        template?: Template,
        mode: ETemplateMode
    }
) => {
    const [tiersToCreate, setTiersToCreate] = useState<Tier[]>([]);
    const [elementsToCreate, setElementsToCreate] = useState<Element[]>([]);
    const [templateName, setTemplateName] = useState<string>('');
    const [saveButtonText, setSaveButtonText] = useState<string>('');
    const [tierHovering, setTierHovering] = useState<number>();
    const [elementHovering, setElementHovering] = useState<number>();

    const theme = useTheme();

    /**
     * Called when in Edit mode to set the existing template informations
     */
    useEffect(() => {
        if (mode === ETemplateMode.Editor && template) {
            setTiersToCreate(template.tiers);
            setElementsToCreate(template.elements);
            setTemplateName(template.name);
            setSaveButtonText('Edit template');
        }

        if (mode === ETemplateMode.Builder) {
            setSaveButtonText('Create template');
        }
    }, [mode]);

    /**
     * Called when a new tier has been created from the Edit view
     * @param {Tier} newTier the tier which was created in the Edit view
     */
    const onTierCreated = (newTier: Tier): void => {
        setTiersToCreate(tiersToCreate.concat(newTier))
    };

    /**
     * Called when a new element has been created from the Edit view
     * @param {Element} newElement the element which was created in the Edit view
     */
    const onElementCreated = (newElement: Element): void => {
        setElementsToCreate(elementsToCreate.concat(newElement))
    };

    /**
     * Called when the save button has been clicked
     */
    const onSaveButtonClick = (): void => {
        const templateToSave: Template = {
            ...template,
            name: templateName,
            tiers: tiersToCreate,
            elements: elementsToCreate
        };

        saveHandler(templateToSave);
    };

    // Called when the template name has changed
    const onNameFieldChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setTemplateName(e.target.value);
    };

    // Called when the delete button of a tier has been clicked
    const onTierDeleteButtonClick = (tierId: number) => {
        setTiersToCreate(tiersToCreate.filter(t => t.id !== tierId));
    };

    // Called when the delete button of an element has been clicked
    const onElementDeleteButtonClick = (elementId: number) => {
        setElementsToCreate(elementsToCreate.filter(e => e.id !== elementId));
    };

    /**
     * Display the list of the created tiers
     * @returns {JSX.Element[]} array of created tiers elements
     */
    const CreatedTiers = (): JSX.Element[] => {
        let createdTiersList: JSX.Element[] = [];
        // Sort tiers by rank
        const sortedTiers: Tier[] = tiersToCreate.sort((a, b) => a.rank - b.rank);

        // Show the delete tier button only on hover
        const DeleteTierButton = (tierId: number): React.JSX.Element | undefined => {
            if (tierHovering === tierId) {
                return (
                    <IconButton edge="end" onClick={() => onTierDeleteButtonClick(tierId)} className={delete_btn_style}>
                        <DeleteIcon />
                    </IconButton>
                );
            }
        };

        sortedTiers.forEach(tier => {
            createdTiersList.push(
                <div
                    className={tier_container_style}
                    style={{backgroundColor: TIERS_COLORS[tier.rank]}}
                    onMouseEnter={() => setTierHovering(tier.id)}
                    onMouseLeave={() => setTierHovering(undefined)}
                >
                    <div className={tier_rank_style}>{tier.rank + 1}</div>
                    <div className={tier_name_style}>{tier.name}</div>
                    <div className="app_spacer"></div>
                    {DeleteTierButton(tier.id!)}
                </div>
            );
        });
        return createdTiersList;
    };

    /**
     * Display the list of the created elements
     * @returns {JSX.Element[]} array of created elements
     */
    const CreatedElements = (): JSX.Element[] => {
        let createdElementsList: JSX.Element[] = [];

        // Show the delete tier button only on hover
        const DeleteElementButton = (elementId: number): React.JSX.Element | undefined => {
            if (elementHovering === elementId) {
                return (
                    <div className={element_delete_container_style}>
                        <RemoveCircleOutlineIcon
                            onClick={() => onElementDeleteButtonClick(elementId)}
                            className={remove_element_icon_style}
                        />
                        <div className={element_delete_bg_style}></div>
                    </div>
                );
            }
        };

        elementsToCreate.forEach(element => {
            createdElementsList.push(
                <div
                    className={element_container_style}
                    onMouseEnter={() => setElementHovering(element.id)}
                    onMouseLeave={() => setElementHovering(undefined)}
                >
                    <img src={TmpElementImg} style={{width: '50px'}} />
                    <div className="app_text_ellipsis">{element.name}</div>
                    {DeleteElementButton(element.id!)}
                </div>
            );
        });
        return createdElementsList;
    };

    return(<>
        <div>
        <TextField
            label="Name"
            variant="outlined"
            onChange={onNameFieldChange}
            value={templateName}
            fullWidth={true}
        />

        <div>Tiers :</div>
        <div>
            {CreatedTiers()}
        </div>

        <TierEditView existingTiers={tiersToCreate} createCallback={onTierCreated} />

        <div>Elements :</div>
        <div className={elements_container_style}>
            {CreatedElements()}
        </div>

        <ElementEditView createCallback={onElementCreated} />
        </div>
        <div className={footer_style}>
            <div className="app_spacer"></div>
            <AppButton text={saveButtonText} onClickHandler={onSaveButtonClick} />
        </div>
    </>);
};

export default TemplateEditor;

/**
 * CSS STYLES
 */
const tier_container_style = css({
    display: 'flex',
    alignItems: 'center',
    width: '100%',
    height: '50px',
    marginTop: '5px'
});

const elements_container_style = css({
    display: 'flex',
    alignItems: 'center'
});

const element_container_style = css({
    display: 'flex',
    alignItems: 'center',
    position: 'relative',
    border: '1px solid',
    flexDirection: 'column',
    width: '100px',
    height: '100px',
    padding: '5px',
    margin: '5px',
    justifyContent: 'center',
    textAlign: 'center'
});

const tier_rank_style = css({
    padding: '0 20px',
    border: '1px solid black',
    height: '100%',
    display: 'flex',
    alignItems: 'center'
});

const tier_name_style = css({
    padding: '0 20px'
});

const delete_btn_style = css({
    margin: '0 15px !important'
});

const element_delete_container_style = css({
    position: 'absolute',
    top: '0',
    left: '0',
    width: '100%',
    height: '100%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    zIndex: 1
});

const element_delete_bg_style = css({
    position: 'absolute',
    top: '0',
    left: '0',
    width: '100%',
    height: '100%',
    opacity: 0.7,
    backgroundColor: 'black',
    zIndex: -1
});

const remove_element_icon_style = css({
    color: 'white',
    width: '65px !important',
    height: '65px !important',
    cursor: 'pointer'
});

const footer_style = css({
    display: 'flex',
    margin: '20px 0'
});
