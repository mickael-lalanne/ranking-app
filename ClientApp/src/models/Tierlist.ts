import { RankedElement } from './Element';
import { UserId } from './User';

export interface Tierlist {
    id: string;
    name: string;
    userId: UserId;
    templateId: string;
    rankedElements: RankedElement[];
    createdAt: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export enum ETierlistDragItem {
    Element = 'element'
}
