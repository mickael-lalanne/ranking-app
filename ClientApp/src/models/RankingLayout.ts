import { Template } from './Template';
import { Tierlist } from './Tierlist';
import TemplatesViewer from '../components/template/TemplatesViewer';
import TemplateEditor from '../components/template/TemplateEditor';
import { createTemplate, deleteTemplate, updateTemplate } from '../services/TemplateServices';
import TierlistsViewer from '../components/tierlist/TierlistsViewer';
import TierlistsEditor from '../components/tierlist/TierlistsEditor';
import { AppDispatch } from '../app/store';
import { createTierlist, deleteTierlist, updateTierlist } from '../services/TierlistServices';

export type RankingType = Template | Tierlist;

export enum ERankingLayoutMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

type EditorComponent = ({ saveHandler, itemToEdit, mode }: EditorComponentProps) => React.JSX.Element;

export interface EditorComponentProps {
    saveHandler: (itemToSave: Template | Tierlist) => Promise<void>;
    itemToEdit?: Template | Tierlist | undefined;
    mode: ERankingLayoutMode;
};

export interface RankingLayoutProps {
    viewerTitle: string;
    viewerSubtitle: string;
    viewerBtnText: string;
    builderTitle: string;
    builderSubtitle: string;
    builderBtnText: string;
    editorTitle: string;
    editorSubtitle: string;
    editorBtnText: string;
    ViewerComponent: ({ editHandler }: {
        editHandler: (template: Template) => void;
    }) => React.JSX.Element;
    EditorComponent: EditorComponent;
    createFunction: (itemToCreate: RankingType, dispatch: AppDispatch) => Promise<any>;
    updateFunction: (itemToUpdate: RankingType, dispatch: AppDispatch) => Promise<any>;
    deleteFunction: (itemIdToDelete: number, dispatch: AppDispatch) => Promise<any>;
}


export const TEMPLATE_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My templates',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add template',
    builderTitle: 'Template creation',
    builderSubtitle: 'Add a name, tiers and elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Template edition',
    editorSubtitle: 'Update your template here',
    editorBtnText: 'Cancel',
    ViewerComponent: TemplatesViewer,
    EditorComponent: TemplateEditor,
    createFunction: createTemplate as (templateToCreate: RankingType, dispatch: AppDispatch) => Promise<Template>,
    updateFunction: updateTemplate as (templateToUpdate: RankingType, dispatch: AppDispatch) => Promise<void>,
    deleteFunction: deleteTemplate
};

export const TIERLIST_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My tierlists',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add tierlist',
    builderTitle: 'Tierlist creation',
    builderSubtitle: 'Rank your elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Tierlist edition',
    editorSubtitle: 'Update your tierlist here',
    editorBtnText: 'Cancel',
    ViewerComponent: TierlistsViewer,
    EditorComponent: TierlistsEditor,
    createFunction: createTierlist as (tierlistToCreate: RankingType) => Promise<Tierlist>,
    updateFunction: updateTierlist as (tierlistToUpdate: RankingType) => Promise<Response>,
    deleteFunction: deleteTierlist
};
