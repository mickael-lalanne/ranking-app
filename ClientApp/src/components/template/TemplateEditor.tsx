import React, { useState, useEffect } from 'react';
import TextField from '@mui/material/TextField';
import TierEditView from './TierEditView';
import { EEditViewMode, TIERS_COLORS, Template, Tier } from '../../models/Template';
import { Element } from '../../models/Element';
import { css } from '@emotion/css';
import ElementEditView from './ElementEditView';
import IconButton from '@mui/material/IconButton';
import DeleteIcon from '@mui/icons-material/Delete';
import AppButton from '../shared/AppButton';
import Button from '@mui/material/Button';
import { ERankingLayoutMode, EditorComponentProps } from '../../models/RankingLayout';
import ElementPreview from '../shared/ElementPreview';
import { useAppDispatch, useAppSelector } from '../../app/hooks';
import { UserId } from '../../models/User';
import { deleteElementsImages, uploadElementsImages } from '../../services/CloudinaryService';
import AddElementButton from '../shared/AddElementButton';
import { ResizedImage, generateRandomId, getTooltipTitleForSaveButtons, isTemporaryId, resizeImage } from '../../services/Util';
import { updateLoading } from '../../store/ApplicationStore';
import { Tooltip } from '@mui/material';

const DEFAULT_TIERS: Tier[] = [
    { id: generateRandomId(), name: 'Tier 1', rank: 0 },
    { id: generateRandomId(), name: 'Tier 2', rank: 1 },
    { id: generateRandomId(), name: 'Tier 2', rank: 2 },
    { id: generateRandomId(), name: 'Tier 4', rank: 3 },
    { id: generateRandomId(), name: 'Tier 5', rank: 4 },
];

const TemplateEditor = (
    { saveHandler, itemToEdit, mode }: EditorComponentProps
) => {
    const [tiersToCreate, setTiersToCreate] = useState<Tier[]>(DEFAULT_TIERS);
    const [elementsToCreate, setElementsToCreate] = useState<Element[]>([]);
    const [elementsImages, setElementsImages] = useState<ResizedImage[]>([]);
    const [templateName, setTemplateName] = useState<string>('');
    const [saveButtonText, setSaveButtonText] = useState<string>('');
    const [tierHovering, setTierHovering] = useState<string>();
    const [editViewMode, setEditViewMode] = useState<EEditViewMode>(EEditViewMode.Hide);
    const [disableSave, setDisableSave] = useState<string>('');

    /**
     * Called when in Edit mode to set the existing template informations
     */
    useEffect(() => {
        if (mode === ERankingLayoutMode.Editor && itemToEdit) {
            setTiersToCreate((itemToEdit as Template).tiers);
            setElementsToCreate((itemToEdit as Template).elements);
            setTemplateName((itemToEdit as Template).name);
            setSaveButtonText('Edit template');
        }

        if (mode === ERankingLayoutMode.Builder) {
            setSaveButtonText('Create template');
        }
    }, [mode]);

    /**
     * Called when a required field value has changed
     * Check if we have to disable the save button or not
     */
    useEffect(() => {
        let disableMessage: string = '';

        if (templateName.length === 0) {
            disableMessage += '- Have a name \n';
        }

        if (tiersToCreate.length === 0) {
            disableMessage += '- Have at least 1 tier \n';
        }

        if (elementsToCreate.length === 0) {
            disableMessage += '- Have at least 1 element \n';
        }

        if (disableMessage.length) {
            disableMessage = 'To be able to save, your template must meet the following requirement(s) : \n' + disableMessage;
        }

        setDisableSave(disableMessage);
    }, [templateName, tiersToCreate, elementsToCreate]);

    const dispatch = useAppDispatch();
    const userId: UserId = useAppSelector(state => state.application.user?.id);
    const loading: boolean = useAppSelector(state => state.application.loading);

    /**
     * Called when a new tier has been created from the Edit view
     * @param {Tier} newTier the tier which was created in the Edit view
     */
    const onTierCreated = (newTier: Tier): void => {
        setTiersToCreate(tiersToCreate.concat(newTier));
        setEditViewMode(EEditViewMode.Hide);
    };

    const onEditViewCancel = () => {
        setEditViewMode(EEditViewMode.Hide);
        setElementsImages([]);
    };

    /**
     * Called when a new element has been created from the Edit view
     * @param {Element[]} newElements the elements which was created in the Edit view
     */
    const onElementCreated = (newElements: Element[]): void => {
        setElementsToCreate(elementsToCreate.concat(newElements));
        setEditViewMode(EEditViewMode.Hide);
    };

    /**
     * Called when the save button has been clicked
     */
    const onSaveButtonClick = async (): Promise<void> => {
        // Show loading indicator
        dispatch(updateLoading(true));

        let elementsWithUploadedImages: Element[] = [];

        // For CREATION, upload ALL the elements images to cloudinary
        if (mode === ERankingLayoutMode.Builder) {
            elementsWithUploadedImages =
                await uploadElementsImages(elementsToCreate, userId!);
        }
        // For EDITION
        else if (mode === ERankingLayoutMode.Editor) {
            // Upload ONLY images that have not already been uploaded
            const elementsAlreadyCreated: Element[] = elementsToCreate.filter(elt => !isTemporaryId(elt.id!));
            const elementsNotYetCreated: Element[] = elementsToCreate.filter(elt => isTemporaryId(elt.id!));

            const newElementsWithImages: Element[] = await uploadElementsImages(elementsNotYetCreated, userId!);

            elementsWithUploadedImages = elementsAlreadyCreated.concat(newElementsWithImages);

            // AND
            // Delete from cloudinary the images that have been removed
            const elementsToDelete: Element[] = (itemToEdit as Template).elements.filter(
                elt => !elementsToCreate.includes(elt)
            );
            await deleteElementsImages(elementsToDelete);
        }

        // Then, save the template in database
        const templateToSave: Template = {
            ...itemToEdit as Template,
            name: templateName,
            tiers: tiersToCreate,
            elements: elementsWithUploadedImages,
            userId
        };

        saveHandler(templateToSave);
    };

    // Called when the template name has changed
    const onNameFieldChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
        setTemplateName(e.target.value);
    };

    // Called when a tier name has been changed
    const onTierNameFieldChange = (
        e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
        tierId: string
    ): void => {
        // Create copy to avoid "Cannot assign to read only property 'name' of object" error
        // It's because tiers have a reference to the redux store
        const tiersCopy: Tier[] = JSON.parse(JSON.stringify(tiersToCreate)); 
        const tierToEditIndex: number = tiersCopy.findIndex(t => t.id === tierId);
        if (tierToEditIndex !== -1) {
            tiersCopy[tierToEditIndex].name = e.target.value;
            setTiersToCreate(tiersCopy);
        }
    };

    // Called when the delete button of a tier has been clicked
    const onTierDeleteButtonClick = (tierId: string) => {
        setTiersToCreate(tiersToCreate.filter(t => t.id !== tierId));
    };

    // Called when the delete button of an element has been clicked
    const onElementDeleteButtonClick = (elementId: string) => {
        setElementsToCreate(elementsToCreate.filter(e => e.id !== elementId));
    };

    // Called when the user has selected one or many images for its elements
    const onElementImageInputChange = async (
        e: React.ChangeEvent<HTMLInputElement>
    ): Promise<void> => {
        if (e.target.files) {
            const resizeFilePromises: Promise<ResizedImage>[] = [];

            // Resize all images selectionned by the user
            Array.from(e.target.files).forEach(file => {
                resizeFilePromises.push(resizeImage(file));
            });

            // Once all images have been resized
            const resizedImages: ResizedImage[] = await Promise.all(resizeFilePromises);

            setElementsImages(resizedImages);
            setEditViewMode(EEditViewMode.EditElement);
        }
    };

    /**
     * Display the list of the created tiers
     * @returns {JSX.Element[]} array of created tiers elements
     */
    const CreatedTiers = (): JSX.Element[] => {
        let createdTiersList: JSX.Element[] = [];
        // Sort tiers by rank
        const sortedTiers: Tier[] = tiersToCreate.slice().sort((a, b) => a.rank - b.rank);

        // Show the delete tier button only on hover
        const DeleteTierButton = (tierId: string): React.JSX.Element | undefined => {
            if (tierHovering === tierId && !loading) {
                return (
                    <IconButton
                        edge="end"
                        onClick={() => onTierDeleteButtonClick(tierId)}
                        className={delete_btn_style}
                        data-cy="delete-tier-button"
                    >
                        <DeleteIcon />
                    </IconButton>
                );
            }
        };

        sortedTiers.forEach(tier => {
            createdTiersList.push(
                <div
                    className={tier_container_style}
                    key={tier.id}
                    style={{backgroundColor: TIERS_COLORS[tier.rank]}}
                    onMouseEnter={() => setTierHovering(tier.id)}
                    onMouseLeave={() => setTierHovering(undefined)}
                    data-cy="template-tier-in-editor"
                >
                    <div className={tier_rank_style}>{tier.rank + 1}</div>
                    <TextField
                        className={tier_name_style}
                        variant="standard"
                        onChange={e => onTierNameFieldChange(e, tier.id!)}
                        value={tier.name}
                        fullWidth
                        disabled={loading}
                        data-cy="tier-name-field"
                    />
                    <div className="app_spacer"></div>
                    {DeleteTierButton(tier.id!)}
                </div>
            );
        });
        return createdTiersList;
    };

    /**
     * Display the list of the created elements
     * @returns {JSX.Element[]} array of created elements
     */
    const CreatedElements = (): JSX.Element[] => {
        let createdElementsList: JSX.Element[] = [];

        elementsToCreate.forEach(element => {
            createdElementsList.push(
                <ElementPreview element={element} key={element.id} deleteElementHandler={onElementDeleteButtonClick} readonly />
            );
        });
        return createdElementsList;
    };

    /**
     * @returns {JSX.Element} the add tier button only if a tier can be created
     */
    const AddTierButton = (): JSX.Element => {
        if (tiersToCreate.length < 5) {
            return (
                <Button
                    variant="contained"
                    className={add_tier_btn_style}
                    onClick={() => setEditViewMode(EEditViewMode.EditTier)}
                    disabled={loading}
                    data-cy="add-tier-button"
                >
                    Add Tier
                </Button>
            );
        }
        return <></>;
    }

    /**
     * @returns {JSX.Element} the add element button if the edit view is not displayed
     */
    const ShowAddElementButton = (): JSX.Element | undefined => {
        if (editViewMode !== EEditViewMode.EditElement) {
            return <div style={{ marginLeft: '5px'}}> 
                <AddElementButton changeCallback={onElementImageInputChange} />
            </div>;
        }
    }

    return(<>
        <div>
            <TextField
                label="Name"
                variant="outlined"
                color="primary"
                data-cy="template-name"
                onChange={onNameFieldChange}
                value={templateName}
                fullWidth={true}
                disabled={loading}
            />

            <div className={editor_title_style}>Tiers :</div>
            <div>
                {CreatedTiers()}
            </div>

            {AddTierButton()}
            <TierEditView
                existingTiers={tiersToCreate}
                createCallback={onTierCreated}
                cancelCallback={onEditViewCancel}
                editViewMode={editViewMode}
            />

            <div className={editor_title_style}>Elements :</div>
            <div className={elements_container_style}>
                {CreatedElements()}
                {ShowAddElementButton()}
            </div>

            <ElementEditView
                createCallback={onElementCreated}
                cancelCallback={onEditViewCancel}
                defaultImages={elementsImages}
                editViewMode={editViewMode}
            />
        </div>
        <div className={footer_style}>
            <div className="app_spacer"></div>
            <Tooltip title={getTooltipTitleForSaveButtons(disableSave)}>
                <div data-cy="create-template">
                    <AppButton
                        text={saveButtonText}
                        onClickHandler={onSaveButtonClick}
                        disabled={loading || !!disableSave}
                    />
                </div>
            </Tooltip>
        </div>
    </>);
};

export default TemplateEditor;

const TIER_HEIGHT:  string = '50px';
const TIER_MARGIN_TOP:  string = '5px !important';
/**
 * CSS STYLES
 */
const tier_container_style = css({
    display: 'flex',
    alignItems: 'center',
    width: '100%',
    height: TIER_HEIGHT,
    marginTop: TIER_MARGIN_TOP
});

const elements_container_style = css({
    display: 'flex',
    alignItems: 'center',
    flexWrap: 'wrap'
});

const tier_rank_style = css({
    padding: '0 20px',
    border: '1px solid black',
    height: '100%',
    display: 'flex',
    alignItems: 'center'
});

const tier_name_style = css({
    padding: '0 20px !important',
    ".MuiInputBase-root:hover:before": {
        borderBottom: '1.5px solid black !important'
    },
    ".MuiInputBase-root:after": {
        borderBottom: '2.2px solid black'
    },
    ".MuiInputBase-root:before": {
        borderBottom: 0
    }
});

const editor_title_style = css({
    margin: '20px 0',
    fontFamily: '"Raleway", sans-serif',
    fontWeight: 300,
    fontSize: '17px'
});

const delete_btn_style = css({
    margin: '0 15px !important'
});

const footer_style = css({
    display: 'flex',
    margin: '20px 0'
});

const add_tier_btn_style = css({
    marginTop: TIER_MARGIN_TOP,
    width: '100%',
    height: TIER_HEIGHT
});
