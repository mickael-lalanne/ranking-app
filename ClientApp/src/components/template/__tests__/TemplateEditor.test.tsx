
import * as React from 'react';
import TemplateEditor from '../TemplateEditor';
import { ERankingLayoutMode } from '../../../models/RankingLayout';
import { renderWithProviders } from '../../../utils/test-utils';

const TEMPLATE_EDITOR_SAMPLE = <TemplateEditor
    saveHandler={ jest.fn()}
    mode={ERankingLayoutMode.Editor}
/>;

describe('Screenshot', () => {
    it('Takes default screenshot', () => {
        const component = renderWithProviders(TEMPLATE_EDITOR_SAMPLE);
    
        expect(component.baseElement).toMatchSnapshot();
    });
});

describe('Common', () => {

    it('Creates a tier', () => {
        // TODO
    });

    it('Creates an element', () => {
        // TODO
    });
    
});

describe('Builder', () => {

});

describe('Editor', () => {

});
