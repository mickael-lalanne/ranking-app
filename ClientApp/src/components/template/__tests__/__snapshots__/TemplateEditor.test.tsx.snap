// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot Takes default screenshot 1`] = `
<body>
  <div>
    <div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        data-cy="template-name"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Name
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="css-ao3svb"
      >
        Tiers :
      </div>
      <div>
        <div
          class="css-f02uu9"
          data-cy="template-tier-in-editor"
          style="background-color: rgb(255, 140, 127);"
        >
          <div
            class="css-7l2k00"
          >
            1
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-y4srn6 css-wb57ya-MuiFormControl-root-MuiTextField-root"
            data-cy="tier-name-field"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                id=":r1:"
                type="text"
                value="Tier 1"
              />
            </div>
          </div>
          <div
            class="app_spacer"
          />
        </div>
        <div
          class="css-f02uu9"
          data-cy="template-tier-in-editor"
          style="background-color: rgb(255, 192, 125);"
        >
          <div
            class="css-7l2k00"
          >
            2
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-y4srn6 css-wb57ya-MuiFormControl-root-MuiTextField-root"
            data-cy="tier-name-field"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                id=":r2:"
                type="text"
                value="Tier 2"
              />
            </div>
          </div>
          <div
            class="app_spacer"
          />
        </div>
        <div
          class="css-f02uu9"
          data-cy="template-tier-in-editor"
          style="background-color: rgb(254, 247, 130);"
        >
          <div
            class="css-7l2k00"
          >
            3
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-y4srn6 css-wb57ya-MuiFormControl-root-MuiTextField-root"
            data-cy="tier-name-field"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                id=":r3:"
                type="text"
                value="Tier 2"
              />
            </div>
          </div>
          <div
            class="app_spacer"
          />
        </div>
        <div
          class="css-f02uu9"
          data-cy="template-tier-in-editor"
          style="background-color: rgb(122, 234, 129);"
        >
          <div
            class="css-7l2k00"
          >
            4
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-y4srn6 css-wb57ya-MuiFormControl-root-MuiTextField-root"
            data-cy="tier-name-field"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                id=":r4:"
                type="text"
                value="Tier 4"
              />
            </div>
          </div>
          <div
            class="app_spacer"
          />
        </div>
        <div
          class="css-f02uu9"
          data-cy="template-tier-in-editor"
          style="background-color: rgb(120, 189, 253);"
        >
          <div
            class="css-7l2k00"
          >
            5
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-y4srn6 css-wb57ya-MuiFormControl-root-MuiTextField-root"
            data-cy="tier-name-field"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                id=":r5:"
                type="text"
                value="Tier 5"
              />
            </div>
          </div>
          <div
            class="app_spacer"
          />
        </div>
      </div>
      <div
        class="css-ao3svb"
      >
        Elements :
      </div>
      <div
        class="css-1dtnjt5"
      >
        <div
          style="margin-left: 5px;"
        >
          <label
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-13ambje css-1t5b4kh css-sghohy-MuiButtonBase-root-MuiButton-root"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AddCircleOutlineIcon"
              focusable="false"
              style="width: 40px; height: 40px;"
              viewBox="0 0 24 24"
            >
              <path
                d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
            <input
              accept="image/*"
              class="css-195j79u"
              data-cy="add-element-input"
              data-testid="add-element-input"
              multiple=""
              type="file"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </label>
        </div>
      </div>
    </div>
    <div
      class="css-9n2jq3"
    >
      <div
        class="app_spacer"
      />
      <div
        class=""
        data-cy="create-template"
        data-mui-internal-clone-element="true"
      >
        <button
          class="
                css-1t5c0s4
                css-av1nc5
            "
          data-cy="app-button"
          disabled=""
          role="button"
        >
          <span
            class="css-1uxuv1w add_btn_shadow"
          />
          <span
            class="css-e5grz5"
          />
          <span
            class="
                    css-1pqqf9o
                    add_btn_front
                    
                "
            style="color: white;"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
